File: CS 201 (Algorithm)/Lecture Qns/Python/DNC Problem Sheet/Number of Line intersections.py
class Solution():
    def intersections(self,P,Q,left,right):
        P.sort()
        return self._intersections(P,Q,left,right)

    def _intersections(self,P,Q,left,right):
        if right - left <= 1:
            return 0

        mid = (left + right)//2

        left_intersections = self.intersections(P, Q, left, mid)
        right_intersections = self.intersections(P,Q,mid,right)

        return left_intersections + right_intersections + self.cross_intersections(P,Q,left,right,mid)

    def cross_intersections(self,P,Q,left,right,mid):
        temp = []
        i = left
        j = mid

        cnt = 0
        while i < mid and j < right:
            if Q[i] <= Q[j]:
                temp.append(Q[i])
                i+=1
            else:
                temp.append(Q[j])
                cnt += mid - i
                j+=1

        while i < mid:
            temp.append(Q[i])
            i += 1

        while j < right:
            temp.append(Q[j])
            j += 1

        Q[left:right] = temp

        return cnt

sol = Solution()
P = [(2,0),(1,0),(4,0),(3,0)]
Q = [(1,1),(4,1),(3,1),(2,1)]

print(sol.intersections(P, Q, 0, len(P)))

File: CS 201 (Algorithm)/Lab Assgn/Problem Sheet DnC 2/Number_of_Line_intersections.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
    public:
        int num_intersections(vector<pair<int,int>> &P, vector<pair<int,int>> &Q) {
            sort(P.begin(),P.end());
            return intersections(P,Q,0,static_cast<int>(P.size()));
        }

        int intersections(vector<pair<int,int>> &P,vector<pair<int,int>> &Q,int left,int right) {
            if (right - left <= 1)
                return 0;

            int mid = (left + right)/2;

            int left_intersections = intersections(P,Q,left,mid);
            int right_intersections = intersections(P,Q,mid,right);

            return left_intersections + right_intersections + cross_intersections(P,Q,left,mid,right);
        }

        int cross_intersections(vector<pair<int,int>> &P, vector<pair<int,int>> &Q,int left,int mid,int right) {
            vector<pair<int,int>> temp;
            int i = left, j = mid;
            int cnt = 0;

            while (i< mid && j < right) {
                if (Q[i] <= Q[j]) {
                    temp.push_back(Q[i]);
                    i++;
                }
                else {
                    cnt += mid - i;
                    temp.push_back(Q[j]);
                    j++;
                }
            }

            while (i < mid) {
                temp.push_back(Q[i]);
                i++;
            }

            while (j<right) {
                temp.push_back(Q[j]);
                j++;
            }


            for (i = 0;i<temp.size();i++) {
                Q[left + i] = temp[i];
            }
            return cnt;
        }
};

int main()
{
    Solution sol;
    vector<pair<int,int>> P = {{2,0},{1,0},{4,0},{3,0}};
    vector<pair<int,int>> Q = {{1,1},{4,1},{3,1},{2,1}};
    cout << sol.num_intersections(P,Q);
    return 0;
};